File: finder4j-backtrack-examples/src/main/java/com/github/K0zka/finder4j/backtrack/examples/sudoku/SudokuStepFactory.java
package com.github.K0zka.finder4j.backtrack.examples.sudoku;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import com.github.K0zka.finder4j.backtrack.StepFactory;

public class SudokuStepFactory implements StepFactory<SudokuStep, SudokuState> {

	private final static class Pair {
		public Pair(short x, short y) {
			super();
			this.x = x;
			this.y = y;
		}

		final short x;
		final short y;

		short getX() {
			return x;
		}

		short getY() {
			return y;
		}
	}

	public List<SudokuStep> produce(SudokuState state) {
		final Pair targetSlot = getLastEmptySlot(state);
		if(targetSlot == null) {
			return Collections.emptyList();
		}
		final ArrayList<SudokuStep> ret = new ArrayList<SudokuStep>();

		final Set<Short> numbersThatCantBe = new HashSet<Short>();
		// add all numbers from the column, the row, and the cell
		for (short x = 0; x < SudokuState.sudokuRowSize; x++) {
			addIfNotNull(numbersThatCantBe,
					state.getSudoku()[x][targetSlot.getY()]);
		}
		for (short y = 0; y < SudokuState.sudokuRowSize; y++) {
			addIfNotNull(numbersThatCantBe,
					state.getSudoku()[targetSlot.getX()][y]);
		}
		final int gridX = targetSlot.getX() - (targetSlot.getX() % SudokuState.sudokuGridSize);
		for (int x = gridX; x < gridX + SudokuState.sudokuGridSize; x++) {
			final int gridY = targetSlot.getY() - (targetSlot.getY() % SudokuState.sudokuGridSize);
			for (int y = gridY; y < gridY + SudokuState.sudokuGridSize; y++) {
				final Short value = state.getSudoku()[x][y];
				addIfNotNull(numbersThatCantBe, value);
			}
		}

		for (short i = 1; i <= SudokuState.sudokuRowSize; i++) {
			if (!numbersThatCantBe.contains(i)) {
				ret.add(new SudokuStep(targetSlot.getX(), targetSlot.getY(), i));
			}
		}

		return ret;
	}

	private void addIfNotNull(final Set<Short> numbersThatCantBe,
			final Short value) {
		if (value != null) {
			numbersThatCantBe.add(value);
		}
	}

	private Pair getLastEmptySlot(SudokuState state) {
		// find first empty spot
		for (short x = 0; x < SudokuState.sudokuRowSize; x++) {
			for (short y = 0; y < SudokuState.sudokuRowSize; y++) {
				if (state.getSudoku()[x][y] == null) {
					return new Pair(x, y);
				}
			}
		}
		return null;
	}

}

File: finder4j-backtrack-examples/src/main/kotlin/com/github/k0zka/finder4j/backtrack/examples/sudoku/SudokuStepFactory.kt
package com.github.k0zka.finder4j.backtrack.examples.sudoku

import com.github.k0zka.finder4j.backtrack.StepFactory
import java.util.ArrayList
import java.util.HashSet

class SudokuStepFactory : StepFactory<SudokuStep, SudokuState> {

	private class Pair(internal val x: Short, internal val y: Short)

	override fun produce(state: SudokuState): List<SudokuStep> {
		val targetSlot = getLastEmptySlot(state) ?: return emptyList<SudokuStep>()
		val ret = ArrayList<SudokuStep>()

		val numbersThatCantBe = HashSet<Short>()
		// add all numbers from the column, the row, and the cell
		for (x in 0 until SudokuState.sudokuRowSize) {
			addIfNotNull(numbersThatCantBe,
						 state.sudoku[x][targetSlot.y.toInt()])
		}
		for (y in 0 until SudokuState.sudokuRowSize) {
			addIfNotNull(numbersThatCantBe,
						 state.sudoku[targetSlot.x.toInt()][y])
		}
		val gridX = targetSlot.x - targetSlot.x % SudokuState.sudokuGridSize
		for (x in gridX until gridX + SudokuState.sudokuGridSize) {
			val gridY = targetSlot.y - targetSlot.y % SudokuState.sudokuGridSize
			for (y in gridY until gridY + SudokuState.sudokuGridSize) {
				val value = state.sudoku[x][y]
				addIfNotNull(numbersThatCantBe, value)
			}
		}

		for (i in 1..SudokuState.sudokuRowSize) {
			if (!numbersThatCantBe.contains(i.toShort())) {
				ret.add(SudokuStep(targetSlot.x, targetSlot.y, i.toShort()))
			}
		}

		return ret
	}

	private fun addIfNotNull(numbersThatCantBe: MutableSet<Short>,
							 value: Short?) {
		if (value != null) {
			numbersThatCantBe.add(value)
		}
	}

	private fun getLastEmptySlot(state: SudokuState): Pair? {
		// find first empty spot
		for (x in 0 until SudokuState.sudokuRowSize) {
			for (y in 0 until SudokuState.sudokuRowSize) {
				if (state.sudoku[x][y] == SudokuState.notSet) {
					return Pair(x.toShort(), y.toShort())
				}
			}
		}
		return null
	}

}
